DocGen4/Output/Structure.lean

```diff
@@ -11,43 +11,54 @@ open Lean
 /--
 Render a single field consisting of its documentation, its name and its type as HTML.
 -/
-def fieldToHtml (f : Process.NameInfo) : HtmlM Html := do
+def fieldToHtml (f : Process.NameInfo) (backrefs : Array BackrefItem) :
```

Comment:
This code is essentially mimicking a state monad, I think it would be better to recall backrefs in `HtmlM`.

Reply:
Could you give me some hints? I'm not familiar with the monadic programming, and stuck with this code in `DocGen4/Output/Base.lean`:
```lean
abbrev HtmlT (m) := ReaderT SiteContext (BaseHtmlT m)
abbrev HtmlM := HtmlT Id
```
The documentation of `ReaderT` says that it only contains a read-only state. I think this should be changed to something which supports writing?

Reply:
You should put a `StateT` transformer on top of it that carries around some `structure SiteState` which contains the backrefs that you are managing. Then provide a couple of monadic helper functions (like the `ReaderT` part already has) to support adding new backrefs and whatever else you need.

For information on monadic programming refer to chapter 5, 6 and in particular 7: https://leanprover.github.io/functional_programming_in_lean/monad-transformers.html I would generally suggest to read through that book before picking up serious programming tasks with Lean.

Reply:
Done.